external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

```diff
@@ -0,0 +1,344 @@
+    while (result.isEmpty && attempt <= MAX_OFFSET_FETCH_ATTEMPTS) {
+      try {
+        result = Some(body)
+      } catch {
+        case e: Exception =>
+          lastException = e
+          logWarning(s"Error in attempt $attempt getting Kafka offsets: ", e)
+          attempt += 1
+          Thread.sleep(OFFSET_FETCH_ATTEMPT_INTERVAL_MS)
+      }
+    }
+    if (result.isEmpty) {
+      assert(attempt > MAX_OFFSET_FETCH_ATTEMPTS)
+      assert(lastException != null)
+      throw lastException
+    }
+    result.get
+  }
+
+  private def positiveMod(a: Long, b: Int): Int = ((a % b).toInt + b) % b
+}
+
+
+/** Companion object for the [[KafkaSource]]. */
+private[kafka010] object KafkaSource {
+
+  val MAX_OFFSET_FETCH_ATTEMPTS = 3
+  val OFFSET_FETCH_ATTEMPT_INTERVAL_MS = 10
+
+  def kafkaSchema: StructType = StructType(Seq(
+    StructField("checksum", LongType),
+    StructField("key", BinaryType),
+    StructField("offset", LongType),
+    StructField("partition", IntegerType),
+    StructField("serializedKeySize", IntegerType),
+    StructField("serializedValueSize", IntegerType),
+    StructField("timestamp", LongType),
+    StructField("timestampType", IntegerType),
+    StructField("topic", StringType),
+    StructField("value", BinaryType)
+  ))
+
+  sealed trait ConsumerStrategy {
+    def createConsumer(): Consumer[Array[Byte], Array[Byte]]
+  }
+
+  case class SubscribeStrategy(topics: Seq[String], kafkaParams: ju.Map[String, Object])
+    extends ConsumerStrategy {
+    override def createConsumer(): Consumer[Array[Byte], Array[Byte]] = {
+      val consumer = new KafkaConsumer[Array[Byte], Array[Byte]](kafkaParams)
+      consumer.subscribe(topics.asJava)
+      consumer.poll(0)
+      consumer
+    }
+
+    override def toString: String = s"Subscribe[${topics.mkString(", ")}]"
+  }
+
+  case class SubscribePatternStrategy(
+    topicPattern: String, kafkaParams: ju.Map[String, Object])
+    extends ConsumerStrategy {
+    override def createConsumer(): Consumer[Array[Byte], Array[Byte]] = {
+      val consumer = new KafkaConsumer[Array[Byte], Array[Byte]](kafkaParams)
+      consumer.subscribe(
+        ju.regex.Pattern.compile(topicPattern),
+        new NoOpConsumerRebalanceListener())
+      consumer.poll(0)
+      consumer
+    }
+
+    override def toString: String = s"SubscribePattern[$topicPattern]"
+  }
+
+  def getSortedExecutorList(sc: SparkContext): Array[String] = {
+    def compare(a: ExecutorCacheTaskLocation, b: ExecutorCacheTaskLocation): Boolean = {
+      if (a.host == b.host) { a.executorId > b.executorId } else { a.host > b.host }
+    }
+
+    val bm = sc.env.blockManager
+    bm.master.getPeers(bm.blockManagerId).toArray
+      .map(x => ExecutorCacheTaskLocation(x.host, x.executorId))
+      .sortWith(compare)
+      .map(_.toString)
+  }
+}
+
+
+/** An [[Offset]] for the [[KafkaSource]]. */
+private[kafka010]
+case class KafkaSourceOffset(partitionToOffsets: Map[TopicPartition, Long]) extends Offset {
```

Comment:
Nit regarding naming -  this is a map of a partition to an offset, not a partition to multiple offsets, at least until we put in 2 starting offsets for newly discovered partitions.